[Clang] [C++26] Expansion Statements (Part 5: Iterating Expansion Statements)

Sirraide opened the pull request:
This implements Sema and template instantiation for iterating expansion statements, as well as expansion statements whose initialiser is dependent (and which could either be iterating or destructuring).

~~One thing that I still need to figure out is how to construct and evaluate a lambda in Sema, because that’s required for computing the expansion size.~~ Update: This is now implemented.

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 5)

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 5)

Sirraide commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169684?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#169690** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169690?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#169689** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169689?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#169688** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169688?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#169687** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169687?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#169686** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169686?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#169685** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169685?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#169684** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169684?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169684?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#169683** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169683?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#169682** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169682?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#169681** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169681?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#169680** <a href="https://app.graphite.com/github/pr/llvm/llvm-project/169680?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

Sirraide commented:
> Update, I think I have figured out the lambda situation at this point (as in building the lambda and evaluating it works for simple examples); now I just need to test it thoroughly.

Ok, just pushed that too and I’ve also implemented support for [CWG 3131](https://github.com/cplusplus/CWG/issues/805), though note that the wording of that issue is likely still subject to change—at least considering the ongoing discussion on the issue. In particular, the comment Barry made wrt the fact that `begin` and `iter` should be `constexpr` only if the for-range-declaration makes sense to me and I’ve implemented that here.

Sirraide commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -56,14 +76,204 @@ static bool HasDependentSize(const CXXExpansionStmtPattern *Pattern) {
+
+  // Try member lookup first.
+  bool FoundBeginEnd = false;
+  if (auto *Record = Ty->getAsCXXRecordDecl()) {
```
Well, this part is just to figure out _whether_ this is an iterating expansion statement; we _do_ perform proper overload resolution and everything below that when we build the actual `begin()`/`end()` calls (irrespective of whether they’re member calls or ADL calls), and that just reuses the for-range part (which is what the refactor in part 4 was for)

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -165,6 +165,11 @@ def err_ice_too_large : Error<
+  "expansion statement size is not a constant expression">;
+
+def note_iterating_expansion_stmt_iterator_requires_minus : Note<
+  "while attempting to construct 'begin - begin' with iterator type %0">;
```
I do not understand this note

Sirraide replied:
Essentially, we need to build the expression `begin - begin` where `begin` is _begin-expr_ as defined by `[stmt.expand]`; I was trying to find some way to notify the user that, if there was an error, it happened while trying to build that expression because the errors can be really confusing otherwise...

cor3ntin commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -11215,6 +11215,12 @@ class Sema final : public SemaBase {
 
+  /// Helper used by the expansion statements and for-range code to build
+  /// a variable declaration for e.g. 'begin' and 'end'. Prefer to use
+  /// BuildCXXForRangeBeginEndVars() instead if applicable.
```
This is an ominous comment - why should i prefer the other function? when is it not applicable?

Sirraide replied:
Yeah, that’s not really a good comment; I’ll remove most of this; it’s fine to use this function on its own (in fact we do in at least one place)

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -20,9 +20,28 @@
+namespace {
+struct IterableExpansionStmtData {
+  enum class State {
+    NotIterable,
```
I think this should either be called `Unknown` or replaced by Matheu's new fangled UnsignedOptional

cor3ntin replied:
Or better yet, have an assert in `TryBuildIterableExpansionStmtInitializer` that we set a state  that we we only need 2 values

Sirraide replied:
I mean, fundamentally there are 3 states here:
1. This is not an iterating expansion statement; we should try destructuring instead.
2. This is iterable, but there was some error trying to build e.g. the declaration of the expansion variable.
3. This is iterable, and there is no error.

I’m not sure using an optional here would make things clearer

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -55,14 +75,204 @@ static bool HasDependentSize(const CXXExpansionStmtPattern *Pattern) {
+    return Data;
+
+  // Lookup member and ADL 'begin()'/'end()'. Only check if they exist; even if
+  // they're deleted, inaccessible, etc., this is still an iterating expansion
+  // statement, albeit an ill-formed one.
+  DeclarationNameInfo BeginName(&S.PP.getIdentifierTable().get("begin"),
+                                ColonLoc);
+  DeclarationNameInfo EndName(&S.PP.getIdentifierTable().get("end"), ColonLoc);
+
+  // Try member lookup first.
+  bool FoundBeginEnd = false;
+  if (auto *Record = Ty->getAsCXXRecordDecl()) {
+    LookupResult BeginLR(S, BeginName, Sema::LookupMemberName);
+    LookupResult EndLR(S, EndName, Sema::LookupMemberName);
+    FoundBeginEnd = S.LookupQualifiedName(BeginLR, Record) &&
+                    S.LookupQualifiedName(EndLR, Record);
+  }
+
+  // Try ADL.
+  //
+  // If overload resolution for 'begin()' *and* 'end()' succeeds (irrespective
+  // of whether it results in a usable candidate), then assume this is an
+  // iterating expansion statement.
```
There are weird things going on with the order and duplication of comments.

Sirraide replied:
Removed the second comment and reworded the third one slightly

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -55,14 +75,204 @@ static bool HasDependentSize(const CXXExpansionStmtPattern *Pattern) {
+
+  auto Ctx = Sema::ExpressionEvaluationContext::PotentiallyEvaluated;
+  if (VarIsConstexpr)
+    Ctx = Sema::ExpressionEvaluationContext::ImmediateFunctionContext;
```
Same question as the other PR

Sirraide replied:
See
- https://github.com/llvm/llvm-project/pull/169681#discussion_r3182730176
- https://github.com/llvm/llvm-project/pull/169682#discussion_r3182774060

tl;dr that helper was meant only for `Parser::ParseForStatement()`

cor3ntin reviewed (commented):
.

cor3ntin commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -242,5 +545,187 @@ Sema::ComputeExpansionSize(CXXExpansionStmtPattern *Expansion) {
+  //    auto e = end-expr;
+  //    for (; b != e; ++b) ++result;
+  //    return result;
+  // }()
```
Oh gosh, this is horrible - did we land consteval blocks yet? Can we avoid code duplication ?

Sirraide replied:
Unfortunately not: consteval blocks (partially) reuse the lambda parsing code... but we have no parser here.

And yes, I agree this is horrible

Endilll commented:
Ambrose, Aaron, and me discussed this during Aaron's office hours in the afternoon. We discussed whether we avoid manually creating a lambda out of thin air in `Sema::ComputeExpansionSize`, which duplicates parser code. An idea was surfaced that we can inject tokens for parser to parse with the usual lambda handling code instead of duplicating it in Sema. Injection can be done via a callback as tokens directly or as a string. My understanding is that Ambrose will explore this approach and will come back to us.

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -165,6 +165,10 @@ def err_ice_too_large : Error<
+def err_expansion_size_expr_not_ice : Error<
+  "expansion statement size is not a constant expression">;
+def err_iterating_expansion_stmt_unsupported : Error<
+  "iterating expansion statements are currently not supported">;
```
```suggestion
  "iterating expansion statements are not yet supported">;
```

shafik commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -165,6 +165,10 @@ def err_ice_too_large : Error<
 def note_constexpr_assert_failed : Note<
   "assertion failed during evaluation of constant expression">;
+def err_expansion_size_expr_not_ice : Error<
+  "expansion statement size is not a constant expression">;
```
Where are these diagnostics tested? Can you comment on each one w/ a link to the PR that tests them? It is really important during review to be able to match a diagnostic w/ a test and I don't want to go hunting around to find it.

Sirraide replied:
All tests (excluding codegen tests) are in patch 7

Sirraide replied:
The vast majority of the Sema tests are in `test/SemaCXX/cxx2c-expansion-stmts.cpp`

shafik commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -11210,6 +11210,11 @@ class Sema final : public SemaBase {
+  /// Helper used by the expansion statements and for-range code to build
+  /// a variable declaration for e.g. 'begin' and 'end'.
+  VarDecl *BuildForRangeVarDecl(SourceLocation Loc, QualType Type,
+                                StringRef Name, bool Constexpr);
```
```suggestion
                                StringRef Name, bool isConstexpr);
```

shafik commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -20,9 +20,28 @@
+namespace {
+struct IterableExpansionStmtData {
+  enum class State {
+    NotIterable,
```
```suggestion
    NonIterable,
```

I think this is better but shrug

shafik commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -20,9 +20,28 @@
 
+namespace {
+struct IterableExpansionStmtData {
+  enum class State {
```
`State` is like used everywhere, can we get a more descriptive name please. Finding where this is used will be painful later on. Please think of the future code spelunkers.

Sirraide replied:
I mean how about `IsIterableResult` or sth like that then? I can’t really think of a ‘good’ name for this...

Sirraide replied:
Changed it to that

shafik commented on clang/lib/Sema/SemaExpand.cpp:
```diff
@@ -20,9 +20,28 @@
+  enum class State {
+    NotIterable,
+    Error,
+    Ok,
```
`NotIterable`, `Error` and `Ok`? This feel like really confusing naming to me, I specially dislike the `isIterable` below which checks `TheState == State::Ok;`

Sirraide replied:
Well the purpose of this is that when we try to determine whether an expansion statement is iterating, the result can be
- it is not iterating (it could still be destructuring; we don’t know that yet);
- there was a hard error (e.g. we determined that it is iterating, but `begin()` is deleted);
- it is iterating, and there was no error.

I’m not sure if there are better names for these enumerators (I’m bad at naming, so I couldn’t think of any); I can add a comment to the enum that explains this though if that would help

Sirraide replied:
Added some comments and renamed `Ok` to `Iterable`

Sirraide committed:
[Clang] [C++26] Expansion Statements (Part 6: Destructuring Expansion Statements) (#169685)

This implements Sema and template instantiation for destructuring
expansion statements.
